### PR TITLE
Add HIP-631 virtual addresses design document

### LINKED ISSUE #4566: HIP 631
### Problem

We need design and test documents for HIP 631 in order to specify the work that needs to be completed.

### Solution

Create a document that specifies the functional and non-functional requirements and technical design for HIP 631. Include required tests cases for acceptance.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by MiroslavGatsanoga**
Signed-off-by: Miroslav Gatsanoga <miroslav.gatsanoga@limechain.tech>

**Description**:
Add design document for account virtual addresses (HIP-631)

**Related issue(s)**:

Fixes #4566

### CODE REVIEW

**Review by stoqnkpL — COMMENTED** [review]
I have some questions related to HIP-631 that may affect the design, so I decided to add them here:
1. Can a `ContractCall` or `ContractCreate` transaction (not `EthereumTransaction`), even if it is signed by an `ECDSA` key, be `recovered` by the EVM via `ecrecover`? What will happen if we call `ecrecover` on a non-Ethereum transaction? How should this behavior be impacted by HIP-631 implementation?
2. Hedera maintains a couple of in-memory maps (after HIP-631 implementation):
    - `evmAddress` → `accountId`
    - `alias` → `accountId`
There are some challenges related to having the implementation depend on in-memory maps:
        - the maps need to be rebuilt on node restart and reconnect
        - executing transactions or queries while the maps are being rebuilt can lead to inconsistent transaction/query results
        - what if in some hard-to-catch corner case, the maps on 2 different nodes are rebuilt differently?
Can we get rid of the in-memory maps and what are the drawbacks of storing these maps in the state?
3. Should the deletion of an account w alias unlink the alias so that in the future another account can use the alias?
4. Should accounts that have at least 1 virtual address be allowed in the EVM if referenced to by their Hedera address? - currently we have a check that prohibits that, e.g. [the cannotUseMirrorAddressOfAliasedContractInPrecompileMethod() EET](https://github.com/hashgraph/hedera-services/blob/d4a6f027e96967a6566224061a460d20f5826e76/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/hapi/ContractCallSuite.java#L399)
5. From HIP-631: "Note: Ethereum address values from known tools (such as Truffle and Hardhat) that have been exposed will be captured by a system account to prevent users from accidentally creating accounts with these breached addresses." 
What are these addresses? Maybe it makes sense to research that and list them here + explain why each address is special.
7. How do we update the account fee logic (to incorporate the new state stored)?

Also, some suggestions for the design document:
1. Can we add a list of global properties related to this HIP, e.g. for the num of Virtual Addresses supported, whether `disables` are supported, etc
2. It may be useful to describe the [phases of development](https://www.notion.so/swirldslabs/Virtual-Address-Roadmap-f6168a668a2a433288248be33e952677) in the design document + impact of each phase, e.g. what protobuf changes will be included in the phase + what fields should be used by transactions, and what behavior impact will the changes have for end-users?

File: docs/design/virtual-addresses.md
Comment (david-bakin-sl) [inline docs/design/virtual-addresses.md:10]: I suggest splitting this into two bullet points so that the maintenance of the map is given more prominence.   (It's also needed in HIP-632.)

File: docs/design/virtual-addresses.md
Comment (david-bakin-sl) [inline docs/design/virtual-addresses.md:12]: What's this about?  More explanation?  (Perhaps in a separate "comments/notes" section elsewhere in this doc?  Or perhaps a pointer to some other doc, maybe a HIP?)

File: docs/design/virtual-addresses.md
Comment (david-bakin-sl) [inline docs/design/virtual-addresses.md:7]: Possibly should be mentioned as an explicit goal: the seamless migration.  (Or is this just understood as a goal?  This is a design doc and migration could be a legitimate goal of its own.)

File: docs/design/virtual-addresses.md
Comment (david-bakin-sl) [inline docs/design/virtual-addresses.md:103]: There should be good tests - pos & neg - specifically relating to the _migration_ (which is not reversible in production).

**Review by david-bakin-sl — COMMENTED** [review]
Plus as discussed in the design review:

- Add reference to the table in the HIP with all the modified transaction record fields
- Add a section on migration issues related to failback - that it's no reversible, what happens if the new data is missing (you look at the old data) etc.

**Comment by david-bakin-sl** [comment]
(Changed PR title to include "HIP-631" to make it easier to find, hope you don't mind)

File: docs/design/virtual-addresses.md
Comment (Nana-EC) [inline docs/design/virtual-addresses.md:33]: Does this class implement an interface or extend another class?
If so might be good to call out the top level methods to be implemented and what they'd handle.

File: docs/design/virtual-addresses.md
Comment (Nana-EC) [inline docs/design/virtual-addresses.md:36]: This is a good ref but If possible please call out expected top level methods to be implemented and how the steps may be grouped.
Would be nice to know how we may plan to tackle them.

File: docs/design/virtual-addresses.md
Comment (Nana-EC) [inline docs/design/virtual-addresses.md:118]: Add migration tests to consider the following entity types.
- simple ED account
- ED account w alias
- simple EC account
- EC account with alias
- hollow account
- contract created via CREATE
- contract created via CREATE2

Also consider a test class to support automated migration testing that verify ledger and in memory maps before and after as an integration test

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Nice work, looking good, some suggestions

File: docs/design/virtual-addresses.md
Comment (Nana-EC) [inline docs/design/virtual-addresses.md:24]: Q: how will account nonce increment be externalized?
If there's no change to this flow for Mirror Node expectations please also call this out.

File: docs/design/virtual-addresses.md
Comment (lukelee-sl) [inline docs/design/virtual-addresses.md:42]: Do we have to move/set the ethereum_nonce

File: docs/design/virtual-addresses.md
Comment (MiroslavGatsanoga) [reply docs/design/virtual-addresses.md]: Updated.

File: docs/design/virtual-addresses.md
Comment (MiroslavGatsanoga) [reply docs/design/virtual-addresses.md]: Added some explanation about what is meant by HIP 32 consistency.

File: docs/design/virtual-addresses.md
Comment (MiroslavGatsanoga) [reply docs/design/virtual-addresses.md:7]: Added seamless migration as an explicit goal.

File: docs/design/virtual-addresses.md
Comment (MiroslavGatsanoga) [reply docs/design/virtual-addresses.md]: Added a comment at the end that it will not be needed initially since it's not a goal.

File: docs/design/virtual-addresses.md
Comment (MiroslavGatsanoga) [reply docs/design/virtual-addresses.md]: Updated to be more explicit with `HederaAccount` interface definition changes.

File: docs/design/virtual-addresses.md
Comment (MiroslavGatsanoga) [reply docs/design/virtual-addresses.md:144]: Updated.

File: docs/design/virtual-addresses.md
Comment (MiroslavGatsanoga) [reply docs/design/virtual-addresses.md]: Updated document with more details about nonces and how they will be externalized.

File: docs/design/virtual-addresses.md
Comment (MiroslavGatsanoga) [reply docs/design/virtual-addresses.md:64]: Yes, I think we should I will be updating the HIP and will add it there in the migration steps.

File: docs/design/virtual-addresses.md
Comment (david-bakin-sl) [reply docs/design/virtual-addresses.md]: (still don't see migration tests in the split-out test-plans/hip-631.md)

File: docs/design/virtual-addresses.md
Comment (MiroslavGatsanoga) [reply docs/design/virtual-addresses.md]: [PR to update HIP](https://github.com/hashgraph/hedera-improvement-proposal/pull/663), also updated class to include nonce related methods.

File: docs/design/virtual-addresses.md
Comment (MiroslavGatsanoga) [reply docs/design/virtual-addresses.md]: Added separate section for migration tests in the test plan doc.

File: docs/design/virtual-addresses.md
Comment (MiroslavGatsanoga) [reply docs/design/virtual-addresses.md]: Added separate section for migration tests in the test plan doc.

File: docs/design/virtual-addresses.md
Comment (lukelee-sl) [inline docs/design/virtual-addresses.md:38]: Will the DefaultVirtualAddress be a separate field `MerkleAccountState`?  If so, how will we ensure that it is in the virtualAddresses set if the virtual address is disabled/removed.

File: docs/design/virtual-addresses.md
Comment (david-bakin-sl) [inline docs/design/virtual-addresses.md:30]: Did this code block get moved here by accident?  Or maybe it belongs here but needs to be separated by some kind of markup - it looks a little odd ...

File: docs/design/virtual-addresses.md
Comment (david-bakin-sl) [inline docs/design/virtual-addresses.md:52]: (lack of mirror node knowledge leads me to ask:) is this map persisted? Or is it always built from state ... when?

File: docs/design/virtual-addresses.md
Comment (lukelee-sl) [inline docs/design/virtual-addresses.md:36]: Why do we need `setVirtualAddresses(Set<ByteString> virtualAddresses)`?  Will not getting the set via `getVirtualAddresses()`suffice to allow us to manipulate the contents of the set (add, delete)?

Also, how will we mark a virtual address as disabled?

File: docs/test-plans/hip-631.md
Comment (MiroslavGatsanoga) [reply docs/test-plans/hip-631.md:62]: Currently no, do you mean to add a test to test something like idempotence of migration logic?

File: docs/design/virtual-addresses.md
Comment (MiroslavGatsanoga) [reply docs/design/virtual-addresses.md:36]: > Why do we need setVirtualAddresses(Set<ByteString> virtualAddresses)? Will not getting the set via getVirtualAddresses()suffice to allow us to manipulate the contents of the set (add, delete)?

I think the idea was to work with immutable collection instead of mutating an underlying one.

> Also, how will we mark a virtual address as disabled?

Hmm, good point. We can extend the map field that we were using for nonces to have values of something like `VirtualAddressSpec`. @Nana-EC any thoughts?

**Comment by hendrikebbers** [comment]
@MiroslavGatsanoga what is the status of this PR. It is inactive for quite some time now.

**Comment by MiroslavGatsanoga** [comment]
@hendrikebbers Anything related to HIP-631 is currently on hold as final updates to HIP-583 are done [here](https://github.com/hashgraph/hedera-improvement-proposal/pull/673).
After HIP-583 updates are merged we will follow up with updating HIP-631 appropriately.

**Comment by Nana-EC** [comment]
@MiroslavGatsanoga let's put this back in draft and decide if it should even be closed until we revisit HIP 631

**Comment by tinker-michaelj** [comment]
Superseded by [HIP-631 itself](https://hips.hedera.com/hip/hip-631).

**Pull request closed without merging**
